packages/ListBox/src/components/Options/helpers/options.js

```diff
@@ -0,0 +1,234 @@
+    }
+  }
+}
+
+export const handleClickOption = ({ props, state, dispatch }) => event => {
+  const { index } = props;
+  const { options, hasFilter, isMulti, refFilterInput } = state;
+  const hasPreventDefaultOnSelect = options[index].preventDefaultOnSelect;
+  if (state.isDisabled || props.isDisabled) {
+    return;
+  }
+
+  if (state.refListBox.current.contains(event.target) && document.activeElement === document.body && !hasFilter) {
+    state.refListBoxContainer.current.focus();
+  }
+
+  if (hasFilter && isMulti) {
+    refFilterInput.current.focus();
+  }
+
+  if (props.onClick || hasPreventDefaultOnSelect) {
+    const onClick = props.onClick ? props.onClick : () => {};
+
+    if (hasPreventDefaultOnSelect) {
+      // this will not selected the option, but will report that was clicked it.
+      // since this action will not affect the state we can report it right back.
+
+      onClick();
+      return;
+    }
+    onClick();
+  }
+
+  if (isMulti) {
+    selectMultipleOption({
+      activeOptionIndex: index,
+      state,
+      dispatch,
+    });
+
+    return;
+  }
+
+  selectSingleOption({ activeOptionIndex: index, isOpen: false, state, dispatch });
+};
+
+export function handleEnterOrSpace({ event, state, dispatch }) {
+  const pressedSpaceKeyWhileHavingFilter = state.hasFilter && event.key === " " && event.target.value !== "";
+
+  const isEventOnFooter = state.refFooterContainer.current
+    ? state.refFooterContainer.current.contains(event.target)
+    : false;
+
+  if (pressedSpaceKeyWhileHavingFilter || isEventOnFooter) {
+    return;
+  }
+
+  event.preventDefault();
+
+  const option = state.options[state.activeOption];
+
+  if (option && (option.isDisabled || option.preventDefaultOnSelect)) {
+    if (state.options[state.activeOption].onClick) {
+      // no sure if this is the state they want
+      // since haven't run the entire cycle befor executing it
+      state.options[state.activeOption].onClick(event, state.activeOption, state.options, state, dispatch);
+    }
+
+    return;
+  }
+
+  if (state.activeOption === null) {
+    if (state.isInline) {
+      return;
+    }
+
+    if (state.noResultsFound) {
+      return;
+    }
+
+    if (state.isMulti) {
+      dispatch({ type: useListBox.types.togglePopover });
+      return;
+    }
+  }
+
+  if (state.isOpen) {
+    if (state.options[state.activeOption].onClick) {
+      state.options[state.activeOption].onClick();
+    }
+
+    if (state.isMulti) {
+      selectMultipleOption({
+        activeOptionIndex: state.activeOption,
+        state,
+        dispatch,
+      });
+      return;
+    }
+
+    // for single select the option is set when the user interact with up and down arrows
+    // no need to notify which option is selected just close the popover
+    dispatch({ type: useListBox.types.closePopover });
+    state.refTrigger.current.focus();
+  } else {
+    dispatch({ type: useListBox.types.openPopover });
+  }
+}
```

Comment:
![smells](https://user-images.githubusercontent.com/14944896/59070387-647a6100-886f-11e9-81a1-717984a17e0f.jpg)

The complexity in here is worrisome, but I suppose this is what you warned us about in your disclaimer:
> Complex logic stuff inside of the component

Hopefully it can be iterated on and improved without any fundamental design changes.

Reply:
Yes, a lot of complexity for managing keyboard. We can keep moving and improving for sure.